[REFACTOR] QnA 리팩토링

gooot opened the pull request:
## What is this PR? 🔍

- 기능 : QnA 리팩토링
- issue : #167 

## Changes 📝

멀티모듈 구조에 맞게 변경

Question
- Read ( Qna )
- Create
- Update
- Delete

Answer
- Create
- Update
- Delete



## Precaution

추후 수정 예정
- Api
- Exception
- Log
- Validation

## ✔️ Please check if the PR fulfills these requirements

- [ ] It's submitted to the correct branch, not the `develop` branch unconditionally?
- [ ] If on a hotfix branch, ensure it targets `main`?
- [ ] There are no warning message when you run `yarn lint`

- Close #167

gooot committed:
refactor : Question 리팩토링

* refactor : Question add 멀티 구조에 맞게 변경

gooot committed:
refactor : Question 리팩토링

* refactor : Question Modify 멀티 구조에 맞게 변경

gooot committed:
refactor : Question 리팩토링

* refactor : Question delete 멀티 구조에 맞게 변경

gooot committed:
refactor : Answer 리팩토링

* refactor : Answer add 멀티 구조에 맞게 변경

gooot committed:
refactor : Answer 리팩토링

* refactor : Answer modify 멀티 구조에 맞게 변경

gooot committed:
refactor : Answer 리팩토링

* refactor : Answer delete 멀티 구조에 맞게 변경

limbaba1120 commented on omocha-api/src/main/java/org/omocha/api/interfaces/dto/QuestionDto.java:
```diff
@@ -0,0 +1,46 @@
+		Long memberId,
+		String email
+
+		//TODO : 유저 정보 수정 필요 ex) profileImageUrl
```
profileImageUrl이랑 nickname도 같이 return을 해야할 것 같습니다

gooot replied:
추가하였습니다.

limbaba1120 commented on omocha-api/src/main/java/org/omocha/api/interfaces/mapper/AnswerDtoMapper.java:
```diff
@@ -0,0 +1,26 @@
+public interface AnswerDtoMapper {
+	AnswerCommand.CreateAnswer toCommand(Long memberId, AnswerDto.CreateAnswerRequest createAnswerRequest);
+
+	AnswerDto.CreateAnswerResponse toDto(AnswerInfo.CreateAnswer createAnswerInfo);
```
toResponse가 더 어울리는 것 같습니다. 아래도 마찬가지

gooot replied:
메소드명 전부 변경 목적에 맞게 변경하였습니다.

limbaba1120 commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+
+		answerValidator.validateAnswerNotExists(question);
+
+		Answer answer = Answer.builder()
```
command에서 entity를 생성하는 method를 만들어서 사용하는게 더 깔끔해보입니다

gooot replied:
모두 Command에서 생성하도록 수정하였습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerReader.java:
QuestionReader & AnswerReader, QuestionStore & AnswerStore는 같이 통합시켜서 QnAReader, QnAStore로 하면 어떨까요?

gooot replied:
통합하였습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerService.java:
Service도 동일하게 하나로 합쳐도 괜찮지 않을까... 싶네요

QnA는 어차피 한 세트로 묶여있기 때문에 합치는게 더 좋아보기이도...

gooot replied:
`QuestionService` 가 100줄
`AuctionService` 가 80줄 정도입니다. 

일단은 그대로 진행하고 다음에 생각해보는게 어떨까요?
Service를 합치게 되면 Command , Info 도 합쳐야 할지...?

ss0ngcode replied:
좋습니다! 일단은 분리해놓고 진행하도록 하죠!

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerValidator.java:
Validator도 합치면 좋을거 같기도 ㅎ...

gooot replied:
우선 Reader와 Store만 합쳐놓았습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/common/TimeTrackableEntity.java:
이거 BaseEntity로 이름 변경되어서 사용중인거 있을거에요!

gooot replied:
BaseEntity 사용하도록 변경하였습니다.

ss0ngcode commented on omocha-infra/src/main/java/org/omocha/infra/repository/AnswerRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface AnswerRepository extends JpaRepository<Answer, Long> {
+	Optional<Answer> findByAnswerIdAndDeletedIsFalse(Long answerId);
```
AnswerRepository이기 때문에 findByIdAndDeletedIsFalse로 해도 될것같아요!

gooot replied:
No property 'id' found for type 'Question'

저는 이렇게 오류가 뜹니다.
혹시 되는거 맞을까요?

ss0ngcode replied:
아 안되나보군요 ㅎㅎ...
method query는 AnswerId가 맞습니다!

ss0ngcode commented on omocha-infra/src/main/java/org/omocha/infra/repository/QuestionRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface QuestionRepository extends JpaRepository<Question, Long> {
+	Optional<Question> findByQuestionIdAndDeletedIsFalse(Long questionId);
```
Question도 위와 동일하게 빠져도 될거같네요!

gooot replied:
No property 'id' found for type 'Question'

저는 이렇게 오류가 뜹니다.
혹시 되는거 맞을까요?

ss0ngcode commented on omocha-infra/src/main/java/org/omocha/infra/AnswerReaderImpl.java:
```diff
@@ -0,0 +1,26 @@
+	@Override
+	public Answer findAnswer(Long answerId) {
+		return answerRepository.findByAnswerIdAndDeletedIsFalse(answerId)
+			.orElseThrow(() -> new NotFoundException("Not found answer"));
```
Exception도 저희가 정한 규칙에 맞게 수정해주세요~

gooot replied:
Exception은 이후 추가하겠습니다.

MemberException을 먼저 만들고 하는게 맞을 것 같습니다.

ss0ngcode commented on omocha-infra/src/main/java/org/omocha/infra/QuestionReaderImpl.java:
```diff
@@ -0,0 +1,26 @@
+
+		// TODO : exception 수정정필요
+		return questionRepository.findByQuestionIdAndDeletedIsFalse(questionId)
+			.orElseThrow(() -> new IllegalArgumentException("Question not found"));
```
여기도 Exception 동일하게 custom으로 만들어서 해주세요!

gooot replied:
Member와 함께 변경하도록 하겠습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+
+		answer.updateAnswer(modifyAnswerCommand.title(), modifyAnswerCommand.content());
+
+		// log.debug("modify answer finished for memberId: {} , answerId: {}, ModifyAnswerRequest : {}", memberId,
+		// 	answerId,
+		// 	modifyAnswerRequest);
```
안쓰는 주석은 삭제해주세요!

gooot replied:
제거하였습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerInfo.java:
```diff
@@ -0,0 +1,49 @@
+		Long questionId,
+		String title,
+		String content,
+		@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
```
info에서는 `@JsonFormat`이 필요 없습니다!

gooot replied:
삭제하였습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerInfo.java:
```diff
@@ -0,0 +1,49 @@
+		Long answerId,
+		String title,
+		String content,
+		@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
+		LocalDateTime createdAt
```
여기도 삭제해주세요!

gooot replied:
삭제하였습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/QuestionInfo.java:
```diff
@@ -0,0 +1,54 @@
+		Long questionId,
+		String title,
+		String content,
+		@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
```
이것도 삭제

gooot replied:
삭제하였습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/QuestionInfo.java:
```diff
@@ -0,0 +1,54 @@
+		Long questionId,
+		String title,
+		String content,
+		@JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
```
여기도 삭제!

gooot replied:
삭제하였습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/QuestionServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+
+		// TODO : Entity 생성 방식 논의 해야함
+		// 		Bid , Auction , Question 다름
+		Question question = Question.builder()
+			.title(createQuestionCommand.title())
+			.content(createQuestionCommand.content())
+			.member(member)
+			.auction(auction)
+			.build();
```
해당 코드는 command로 옮기거나 store에서 처리하도록 해주세요!
여기는 command가 존재하니깐 command 내부에서 처리하는게 좋겠네요

gooot replied:
Command에서 처리하도록 하였습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/QuestionServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+
+		question.updateQuestion(modifyQuestionCommand.title(), modifyQuestionCommand.content());
+
+		// log.debug("modify question finished for memberId: {}, questionId: {}, ModifyQuestionRequest: {}", memberId,
+		// 	questionId, modifyQuestionRequest);
```
안쓰는 주석 삭제

gooot replied:
삭제하였습니다.

ss0ngcode reviewed (commented):
리뷰 남겼습니다! 확인해주세요!

gooot committed:
merge commit

gooot committed:
refactor : QnA 통합 조회 추가

* refactor : 멀티 모듈 구조에 맞게 변경

gooot committed:
refactor : PR 리뷰 반영 , 컨벤션 적용

* refactor : command 에서 Entity 생성

* refactor : method명 수정
- 컨벤션 적용

* refactor : QuestionReader , AnswerReader
find , get 분리

* refactor : QnaReader 통합
- QuestionReader.java
- AnswerReader.java

* refactor : QnaStore 통합
- QuestionStore.java
- AnswerStore.java

* refactor : Exception , validator 임시 추가

gooot committed:
refactor : 누락 수정

* refactor : AnswerRepository 미사용 코드 삭제

* refactor : AnswerValidatorImpl
- Reader 변경 사항 적용

limbaba1120 commented on omocha-api/src/main/java/org/omocha/api/interfaces/dto/QuestionDto.java:
```diff
@@ -0,0 +1,59 @@
+	) {
+	}
+
+	public record QuestionDetails(
```
이거는 어디에 사용이 되나요? 그리고 Request or Response 필요할 것 같아요

gooot replied:
`QuestionDetails` , `AnswerDetails` 는 따로 response 로 반환이 되지는 않고 

`RetriveQnasResponse ` 안에 위치해 있습니다.

저렇게 만든 목적은 프론트에서 [question , answer] 이런 식으로 짝지어져서 오는게 더 편하다고 해서 만들었습니다.

limbaba1120 commented on omocha-api/src/main/java/org/omocha/api/interfaces/dto/QuestionDto.java:
```diff
@@ -0,0 +1,59 @@
+	) {
+	}
+
+	public record AddQuestionResponse(
```
Controller명 이랑 동일해야할 것 같아요. QuestionAddResponse, QuestionAddRequest

limbaba1120 commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerInfo.java:
```diff
@@ -0,0 +1,70 @@
+
+public class AnswerInfo {
+
+	public record AnswerDetails(
```
이 부분은 service,facade명과 동일해야할 것 같아요

limbaba1120 commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+	public AnswerInfo.AddAnswer addAnswer(AnswerCommand.AddAnswer addAnswerCommand) {
+		log.info("add answer started for createAnswerCommand: {}", addAnswerCommand);
+
+		// TODO : Entity 조회 추후 리팩토링
```
Refactoring 된건가요? 그러면 TODO는 삭제해주세요. 그리고 위에 log는 controoller에 적어주세요

limbaba1120 commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/Qna.java:
```diff
@@ -0,0 +1,21 @@
+
+@Data
+@NoArgsConstructor
+public class Qna {
```
이거는 뭔가요? Entity인가요?

gooot replied:
해당 코드부분 코드리뷰 받은 내용입니다.

Q4. QueryDSL
이 경우에는 Response가 아니라 두 테이블을 조인한 새로운 Domain이라고 보는편이 맞겟네요. 그렇다면 Domain에 위치 하는게 맞지 않을까요?

이렇게 말씀하셔서 우선 `도메인` 이라고 생각하고 넣었습니다.

개인적으로 이 부분은 좀 더 개선 할 수 있다고 보고 있습니다. 
하지만 계속 붙잡고 있으면 시간이 오래 소요 될 것 같아서 
이후 따로 시간을 투자해서 개선해보려고 합니다.

limbaba1120 commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerValidatorImpl.java:
```diff
@@ -0,0 +1,37 @@
+	public void validateAnswerNotExists(
+		Question question
+	) {
+		if (qnaReader.existsByQuestionId(question.getQuestionId())) {
```
existByQuestId 보다 명확한 method명이 필요할 것 같아요.QuestionId로 Question이 있는 지 확인하는 코드인 줄 알았습니다.

answerExist 라던지 다른걸로 변경이 필요할 것 같아요!

ss0ngcode replied:
checkAnswerExist 정도가 좋을 수 있겠네요~

gooot replied:
2개의 Reader를 하나로 합치면서 의미가 모호해 진 것 같습니다.

원래의 의미는 AnswerRepository에 답변이 있는지 questionId 로 확인 하는 코드입니다.

gooot replied:
수정하였습니다.

limbaba1120 commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/QuestionServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+
+	@Override
+	@Transactional(readOnly = true)
+	public Page<QuestionInfo.RetriveQnas> retriveQnas(QuestionCommand.QnaList qnaListCommand,
+		Pageable sortPage) {
```
retrieve 입니다~

limbaba1120 commented on omocha-infra/src/main/java/org/omocha/infra/repository/QuestionRepositoryCustom.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.data.domain.Pageable;
+
+public interface QuestionRepositoryCustom {
+	Page<Qna> retribeQnaList(
```
retrieve

limbaba1120 commented on omocha-infra/src/main/java/org/omocha/infra/repository/QuestionRepositoryImpl.java:
```diff
@@ -0,0 +1,68 @@
+	}
+
+	@Override
+	public Page<Qna> retribeQnaList(Long auctionId, Pageable pageable) {
```
retrieve

ss0ngcode commented on omocha-api/src/main/java/org/omocha/api/interfaces/QuestionController.java:
```diff
@@ -0,0 +1,164 @@
+	private final QuestionDtoMapper questionDtoMapper;
+	private final PageSort pageSort;
+
+	// TODO : QueryDSL JOIN 관련 수정 필요
+	@GetMapping("/{auctionId}/qna-list")
+	public ResponseEntity<ResultDto<Page<QuestionDto.RetriveQnasResponse>>> qnaList(
+		@PathVariable(value = "auctionId") Long auctionId,
+		@RequestParam(value = "sort", defaultValue = "createdAt") String sort,
+		@RequestParam(value = "direction", defaultValue = "ASC") String direction,
+		@PageableDefault(page = 0, size = 10)
+		Pageable pageable
+	) {
+
+		// TODO : PR 전 TODO 수정해야됨
```
TODO가 처리 된 내용들일까요?
확인 한 번만 부탁드려요!

ss0ngcode replied:
direction 이것도 최신순이 아닌 오래된순으로 먼저 보여주는게 default가 맞을까요?

gooot replied:
QnA에 질문이 많이 달릴거라고 생각은 안하지만 많이 달리게 된다면

`최신순`으로 보여주게 된다면
다음 페이지로 넘어가는 과정에서 size 만큼 글이 추가된다면 같은 페이지를 보여주게 됩니다. 
이런 문제 때문에 `오래된순`으로 보여주는게 맞지 않을까?라고 생각이 됩니다.

ss0ngcode commented on omocha-api/src/main/java/org/omocha/api/interfaces/QuestionController.java:
```diff
@@ -0,0 +1,164 @@
+
+		Pageable sortPage = pageSort.sortPage(pageable, sort, direction);
+
+		QuestionCommand.QnaList qnaListCommand = questionDtoMapper.toCommand(auctionId);
```
command에 값이 하나밖에 없고 facade와 service에서 그 값만 사용한다면 mapper로 변환하지 않고 그냥 auctionId를 Long으로 넘겨주는것도 좋을 것 같아요~
그런데 sortPage값이 있다면 그걸 command로 같이 변환해주는것도 좋겠네요!

ss0ngcode replied:
이건 취향차이라 선호하시는 방법으로 하셔도 무방할 것 같습니다~

gooot replied:
저도 하나라서 Long으로 넘겨도 될거 같다고 생각했습니다만 
auction에서도 1개라도 넘기는 부분이 있어서 통일하려고 넣었습니다. 

추후 수정할 부분이 생긴다면 field를 추가하면 되니까 그런 점도 좋다고 생각했습니다.

ss0ngcode commented on omocha-api/src/main/java/org/omocha/api/interfaces/mapper/QuestionDtoMapper.java:
```diff
@@ -0,0 +1,47 @@
+)
+public interface QuestionDtoMapper {
+
+	QuestionCommand.ModifyQuestion toCommand(Long memberId, Long questionId,
+		QuestionDto.ModifyQuestionRequest modifyQuestionRequest);
```
이것도 아래처럼 개행을 해주시면 보기 좋을 것 같네요!
```suggestion
	QuestionCommand.ModifyQuestion toCommand(
        Long memberId, 
        Long questionId,
		QuestionDto.ModifyQuestionRequest modifyQuestionRequest
    );
```

ss0ngcode commented on omocha-api/src/main/java/org/omocha/api/interfaces/mapper/AnswerDtoMapper.java:
```diff
@@ -0,0 +1,28 @@
+
+	AnswerDto.AddAnswerResponse toResponse(AnswerInfo.AddAnswer createAnswerInfo);
+
+	AnswerCommand.ModifyAnswer toCommand(Long memberId, Long answerId,
+		AnswerDto.ModifyAnswerRequest modifyAnswerRequest);
```
이것도 개행!
```suggestion
	AnswerCommand.ModifyAnswer toCommand(
        Long memberId, 
        Long answerId,
		AnswerDto.ModifyAnswerRequest modifyAnswerRequest
    );
```

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/Answer.java:
```diff
@@ -0,0 +1,68 @@
+
+	// TODO : SoftDelete 논의 후 결정
+	//  field 추가 or removedAt 추가
+	@Column(name = "deleted", nullable = false)
+	@ColumnDefault("false")
+	private boolean deleted;
```
일단 이건 주석처리 해주세요
SoftDelete 논의하고 다시 정하도록 하죠!
TODO는 그대로 남겨주시구요!

gooot replied:
해당 부분은 우선 그대로 두는 것이 좋아보입니다.

이유 
1. 주석 처리 시 CRUD에도 수정사항을 반영해야해서 수정 범위가 넓습니다.
2. SoftDelete 회의 후 한번에 수정하는 것이 좋아보입니다. 회의에 따라 유지 될 수도 있는 부분이고 현재 수정을 하게된다면 2번 수정을하게 됩니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/AnswerValidatorImpl.java:
```diff
@@ -0,0 +1,37 @@
+	}
+
+	@Override
+	public void validateAnswerNotExists(
```
이 메서드명도 checkAnswerExists로 하면 어떨까요?

gooot replied:
validateAnswerNotExists 이 맞는 것 같습니다.

해당 질문에 답변이 존재한다면 답변을 추가할 수 없습니다.

validator 에서 검증하는 부분에 대해서 
해당 메소드에 대한 답이 긍정이면 항상 정상 로직을 나타내도록 만들고 싶었습니다.

ss0ngcode commented on omocha-domain/src/main/java/org/omocha/domain/auction/qna/QuestionValidatorImpl.java:
```diff
@@ -0,0 +1,38 @@
+	public final QnaReader qnaReader;
+
+	@Override
+	public void validModifyAndRemove(Question question) {
```
이것도 메서드명을 checkAnswerExists로 하면 어떨까 싶은데...
개인적인 의견이라 뭐가 맞는지는 고민이네요 ㅎㅎ...

ss0ngcode reviewed (commented):
리뷰 체크 부탁드려요!

gooot committed:
refactor : PR 반영

* refactor : DTO , Command , Info 컨벤션 적용

* refactor : method 컨벤션 적용

* refactor : controller log 추가

* refactor : validator 메소드명 변경

gooot committed:
refactor : PR 반영 누락

ss0ngcode reviewed (approved):
LGTM!
